### PR TITLE
More exception handling, less assertions

### LINKED ISSUE #979: A concrete example of moving from assertions to exceptions
Hi! I'm the lead maintainer of the Data Carpentry [Python for Atmosphere and Ocean Science](https://carpentries-lab.github.io/python-aos-lesson/) (PyAOS) lessons.

When I first wrote the PyAOS lessons a few years ago, I borrowed heavily from the Software Carpentry lessons. My defensive programming lesson was almost a straight copy of the corresponding python-novice-inflammation lesson. However, last year I received feedback (https://github.com/carpentries-lab/python-aos-lesson/issues/36) similar to #622, #773 and #874, suggesting that in most cases raising exceptions is a better approach to using assertions. I therefore just went through the process of re-writing our defensive programming lesson to address that feedback:  

https://carpentries-lab.github.io/python-aos-lesson/08-defensive/index.html

Instead of teaching assertions and then test-driven development, we now teach the following (quoted directly from the lesson):

_The first step toward getting the right answers from our programs is to assume that mistakes will happen and to guard against them. This is called defensive programming, and there are a number of tools and approaches at our disposal for doing this. Broadly speaking, we can:_
- _raise and handle errors to check and respond to program inputs_
- _use assertions to make sure nothing crazy or unexpected has happened_
- _write unit tests to make sure each component of our program produces expected outputs_
- _use a logging framework to report on program activity_

_In this lesson, we’ll look at how error handling, assertions and logging can make the unit conversion in our program more reliable, and we’ll provide links to further information on unit testing._

If people think our new PyAOS defensive programming lesson looks like a good approach to addressing the concerns in #622, #773 and #874, I'd be happy to submit a PR that adapts the content in our PyAOS lesson to fit with the Software Carpentry defensive programming lesson?

### PR DESCRIPTION
**Pull request by DamienIrving**
Resolves #979.

This is a substantial re-write of the defensive programming lesson due to lots of feedback in recent years pointing out that exception handling is in many/most cases a better alternative to assertions.

The re-write proposed in this PR can be view at:  
https://damienirving.github.io/python-novice-inflammation/10-defensive/index.html

### CODE REVIEW

**Comment by DamienIrving** [comment]
This is now ready for review. 

I've trimmed back the assertion content and added new content on exception handling (as was the brief for this PR). This means I've left the test driven development section essentially unchanged. If people feel that the lesson is now too long the most obvious place to shorten the lesson would be that section - the `range_overlap` example is quite long. By writing five assertions and then putting them in a `test_range_overlap` function we're also super close to doing what you'd need to do in order to run pytest (you'd just put each assertion in its own function and put those five functions in a script called `test_range_overlap.py`), so we could also consider actually using a testing framework (i.e. pytest) rather than just mentioning they exist.

File: _episodes/10-defensive.md [hunk 1]
```diff
@@ -47,225 +49,229 @@ As in real carpentry --- the kind done with lumber --- the time saved
...
-AssertionError: Rectangles must contain 4 coordinates
+ValueError: The low_val is >= the high_val
 ~~~
 {: .error}
 
+## Handling errors
+
+As we've seen in the examples above and in the previous lesson,
+if exceptions aren't dealt with the program crashes.
+The error message upon crashing is sometimes be easy to understand
```
Comment (tobyhodges) [inline _episodes/10-defensive.md:156]:
```suggestion
The error message upon crashing is sometimes easy to understand
```

File: _episodes/10-defensive.md [hunk 2]
```diff
@@ -47,225 +49,229 @@ As in real carpentry --- the kind done with lumber --- the time saved
...
-10-20% of the code they contain
-are there to check that the other 80–90% are working correctly.
-Broadly speaking,
-assertions fall into three categories:
-
-*   A [precondition]({{ page.root }}/reference.html#precondition)
-    is something that must be true at the start of a function in order for it to work correctly.
+In this case we've chosen to raise a generic `Exception`,
+but we could pick any of the [builtin exception types](https://docs.python.org/3/library/exceptions.html#exception-hierarchy)
+or define our own [custom exception class](https://towardsdatascience.com/how-to-define-custom-exception-classes-in-python-bfa346629bca).
```
Comment (tobyhodges) [inline _episodes/10-defensive.md:77]: I would emphasise here that understanding how to define custom exception classes requires knowledge of concepts that are beyond the scope of this lesson.

**Review by tobyhodges — COMMENTED** [review]
I have left a couple comments but would like to see @swcarpentry/curriculum-advisors review this pull request before it is merged. 

(The CAC has not met for a few months, while they waited for a new secretary to join the group. Now that is done, I hope you will be able to get some feedback on this soon @DamienIrving.)

File: _episodes/10-defensive.md (hunk 1 above)
Comment (DamienIrving) [reply _episodes/10-defensive.md]: Fixed :smile:

File: _episodes/10-defensive.md (hunk 2 above)
Comment (DamienIrving) [reply _episodes/10-defensive.md]: Fixed :smile:

**Comment by bsmith89** [comment]
> I have left a couple comments but would like to see @swcarpentry/curriculum-advisors review this pull request before it is merged.
> 
> (The CAC has not met for a few months, while they waited for a new secretary to join the group. Now that is done, I hope you will be able to get some feedback on this soon @DamienIrving.)

I'll add this topic to the agenda for our next meeting. :)

**Comment by bsmith89** [comment]
Existing relevant discussion "issue" in our repo: swcarpentry/curriculum-advisors#1

**Comment by noatgnu** [comment]
Hi @DamienIrving ,

Thank you for the pull requests. The curriculum advisors are split on this so we will close this request in the mean time to be open at a later time after migration to Workbench.

**Comment by zkamvar** [comment]
Hello @DamienIrving, 

Thank you for your work on this. I am the lesson infrastructure developer and one of the impacts of the transition to The Workbench by May is that all open pull requests must be merged and the forks re-created (https://carpentries.github.io/workbench/faq.html#github-forks). 

Because your contribution is substantial, if you are not interested in starting the re-submission from scratch, I can help you transition your work to use The Workbench so that, after May, you can re-contribute and build from where you left off. 

Sincerely,
Zhian

**Comment by DamienIrving** [comment]
Hi @zkamvar,

If you could help me transition my work to use The Workbench so I can re-contribute that would be great. Where do I start?

**Comment by DamienIrving** [comment]
Apologies for the delay in responding, @zkamvar. Option 1 sounds good to me. How should we proceed with that option?

**Pull request closed without merging**